README.md
# timestring

[![Version](https://img.shields.io/npm/v/timestring.svg?style=flat-square)](https://www.npmjs.com/package/timestring)
[![Build Status](https://img.shields.io/github/actions/workflow/status/mike182uk/timestring/ci.yml?branch=main&style=flat-square)](https://github.com/mike182uk/timestring/actions/workflows/ci.yml?query=workflow%3ACI)
[![Coveralls](https://img.shields.io/coveralls/mike182uk/timestring/main.svg?style=flat-square)](https://coveralls.io/r/mike182uk/timestring)
[![npm](https://img.shields.io/npm/dm/timestring.svg?style=flat-square)](https://www.npmjs.com/package/timestring)
[![License](https://img.shields.io/github/license/mike182uk/timestring.svg?style=flat-square)](https://www.npmjs.com/package/timestring)

Parse a human readable time string into a time based value.

## Installation

```bash
npm install --save timestring
```

## Usage

### Overview

```js
const timestring = require('timestring')

let str = '1h 15m'
let time = timestring(str)

console.log(time) // will log 4500
```

**By default the returned time value from `timestring` will be in seconds.**

The time string can contain as many time groups as needed:

```js
const timestring = require('timestring')

let str = '1d 3h 25m 18s'
let time = timestring(str)

console.log(time) // will log 98718
```

and can be as messy as you like:

```js
const timestring = require('timestring')

let str = '1 d    3HOurS 25              min         1   8s'
let time = timestring(str)

console.log(time) // will log 98718
```

### Keywords

`timestring` will parse the following keywords into time values:

1. `ms, milli, millisecond, milliseconds` - will parse to milliseconds
2. `s, sec, secs, second, seconds` - will parse to seconds
3. `m, min, mins, minute, minutes` - will parse to minutes
4. `h, hr, hrs, hour, hours` - will parse to hours
5. `d, day, days` - will parse to days
6. `w, week, weeks` - will parse to weeks
7. `mon, mth, mths, month, months` - will parse to months
8. `y, yr, yrs, year, years` - will parse to years

Keywords can be used interchangeably:

```js
const timestring = require('timestring')

let str = '1day 15h 20minutes 15s'
let time = timestring(str)

console.log(time) // will log 141615
```

### Return Time Value

By default the return time value will be in seconds. This can be changed by passing one of the following strings as an argument to `timestring`:

1. `ms` - Milliseconds
2. `s` - Seconds
3. `m` - Minutes
4. `h` - Hours
5. `d` - Days
6. `w` - Weeks
7. `mth` - Months
8. `y` - Years

```js
const timestring = require('timestring')

let str = '22h 16m'

let hours = timestring(str, 'h')
let days = timestring(str, 'd')
let weeks = timestring(str, 'w')

console.log(hours) // will log 22.266666666666666
console.log(days) // will log 0.9277777777777778
console.log(weeks) // will log 0.13253968253968254
```

### Optional Configuration

A few assumptions are made by default:

1. There are 24 hours per day
2. There are 7 days per week
3. There are 4 weeks per month
4. There are 12 months per year
5. There are 365.25 days per year

These options can be changed by passing an options object as an argument to `timestring`.

The following options are configurable:

1. `hoursPerDay`
2. `daysPerWeek`
3. `weeksPerMonth`
4. `monthsPerYear`
5. `daysPerYear`

```js
const timestring = require('timestring')

let str = '1d'
let opts = {
  hoursPerDay: 1
}

let time = timestring(str, 'h', opts)

console.log(time) // will log 1
```

In the example above `hoursPerDay` is being set to `1`. When the time string is being parsed, the return value is being specified as hours. Normally `1d` would parse to `24` hours (as by default there are 24 hours in a day) but because `hoursPerDay` has been set to `1`, `1d` will now only parse to `1` hour.

This would be useful for specific application needs.

*Example - Employees of my company work 7.5 hours a day, and only work 5 days a week. In my time tracking app, when they type `1d` i want 7.5 hours to be tracked. When they type `1w` i want 5 days to be tracked etc.*

```js
const timestring = require('timestring')

let opts = {
  hoursPerDay: 7.5,
  daysPerWeek: 5
}

let hoursToday = timestring('1d', 'h', opts)
let daysThisWeek = timestring('1w', 'd', opts)

console.log(hoursToday) // will log 7.5
console.log(daysThisWeek) // will log 5
```

It is important to note that the `daysPerYear` configuration option will be used to convert a month or year to seconds, so if you are using custom configuration options make sure that you adjust this value to suit if you expect to be parsing timestrings containing months or years.

## Notes

If the string that is passed into `timestring` can not be parsed then an error will be thrown:

```js
const timestring = require('timestring')

let str = 'aaabbbccc'
let time = timestring(str) // will throw an error
```

If a number is passed into `timestring` it will be treated as a string containing milliseconds:

```js
const timestring = require('timestring')

let time = timestring(3000) // 3s
```


CHANGELOG.md
# Changelog

## 7.0.0

- Drop support for Node.js `< 12.0.0`
- If a numeric value is provided it is trated as a string containing milliseconds ([alexeymarunin](https://github.com/alexeymarunin))

## 6.0.0

- Drop support for Node.js `< 8.0.0`
- Throw an error when unable to parse a timestring ([prolink007](https://github.com/prolink007))

## 5.0.1

- Fix issue where decimal values were not being parsed correctly ([vegeta897](https://github.com/vegeta897))

## 5.0.0

- Add `daysPerYear` configuration option
- Use `daysPerYear` configuration option to convert months or years to seconds

## 4.0.0

- Drop support for Node.js `< 4.0.0`
- Remove dependency on `lodash`

## 3.2.0

- Add `mon` unit ([@IanMitchell](https://github.com/IanMitchell))

## 3.1.0

- Add `ms` unit ([@mrbar42](https://github.com/mrbar42))

## 3.0.0

- Remove `String.parseTime`
- Export function instead of object

## 2.0.0

- Keep it simple: revert back to ES5, remove gulp, browserify and build steps
- Remove bower / browser specific integrations - This module can still be used client side using modern development tools like webpack, browserify etc.
- use `node-style-guide` coding style

## 1.1.1

- src now uses ES6
- use gulp instead of grunt
- setup gulp build pipeline
- use analysis tools: JSCS, jshint
- use consistent coding style (jQuery coding style)
- send code coverage to coveralls.io
- use browserify to generate distributable scripts
- add contributing doc
- misc meta data updates
- add sourcemaps

## 1.1.0

- add MIT license
- add Changelog
- add Grunt config
- add editorconfig
- add Tests
- fix JSHint errors
- format code as per editorconfig
- fix bug with conversion from year(s) to another unit
- make sure `Error` (type) is thrown when invalid unit is encountered
- add minified distributable version
- update package.json
- add bower config
- add travis config


test.js
/* globals describe, it */

const { expect } = require('chai')
const timestring = require('./index')

describe('timestring', () => {
  it('can parse a timestring', () => {
    expect(timestring('500ms')).to.equal(0.5)
    expect(timestring('1s')).to.equal(1)
    expect(timestring('1m')).to.equal(60)
    expect(timestring('1h')).to.equal(3600)
    expect(timestring('1d')).to.equal(86400)
    expect(timestring('1w')).to.equal(604800)
    expect(timestring('1mth')).to.equal(2629800)
    expect(timestring('1y')).to.equal(31557600)
  })

  it('can parse different unit identifiers', () => {
    const unitMap = {
      ms: ['ms', 'milli', 'millisecond', 'milliseconds'],
      s: ['s', 'sec', 'secs', 'second', 'seconds'],
      m: ['m', 'min', 'mins', 'minute', 'minutes'],
      h: ['h', 'hr', 'hrs', 'hour', 'hours'],
      d: ['d', 'day', 'days'],
      w: ['w', 'week', 'weeks'],
      mth: ['mon', 'mth', 'mths', 'month', 'months'],
      y: ['y', 'yr', 'yrs', 'year', 'years']
    }

    unitMap.ms.forEach(unit => {
      expect(timestring(`500 ${unit}`)).to.equal(0.5)
    })

    unitMap.s.forEach(unit => {
      expect(timestring(`3 ${unit}`)).to.equal(3)
    })

    unitMap.m.forEach(unit => {
      expect(timestring(`2 ${unit}`)).to.equal(120)
    })

    unitMap.h.forEach(unit => {
      expect(timestring(`7 ${unit}`)).to.equal(25200)
    })

    unitMap.d.forEach(unit => {
      expect(timestring(`4 ${unit}`)).to.equal(345600)
    })

    unitMap.w.forEach(unit => {
      expect(timestring(`2 ${unit}`)).to.equal(1209600)
    })

    unitMap.mth.forEach(unit => {
      expect(timestring(`9 ${unit}`)).to.equal(23668200)
    })

    unitMap.y.forEach(unit => {
      expect(timestring(`1 ${unit}`)).to.equal(31557600)
    })
  })

  it('can return a value in a specified unit', () => {
    expect(timestring('1m', 'ms')).to.equal(60000)
    expect(timestring('5m', 's')).to.equal(300)
    expect(timestring('5m', 'm')).to.equal(5)
  })

  it('uses the passed options instead of the defaults', () => {
    const opts = {
      hoursPerDay: 1,
      daysPerWeek: 2,
      weeksPerMonth: 3,
      monthsPerYear: 4,
      daysPerYear: 30
    }

    expect(timestring('1d', 'h', opts)).to.equal(1)
    expect(timestring('1w', 'd', opts)).to.equal(2)
    expect(timestring('1mth', 'w', opts)).to.equal(3.75)
    expect(timestring('1y', 'mth', opts)).to.equal(4)
  })

  it('throws an error when an invalid unit is used in the timestring', () => {
    expect(() => timestring('1g')).to.throw(Error)
  })

  it('throws an error when no numbers are in the timestring', () => {
    expect(() => timestring('asdf')).to.throw(Error)
  })

  it('throws an error when numbers tail the timestring', () => {
    expect(() => timestring('asdf123')).to.throw(Error)
  })

  it('can parse a messy time string', () => {
    expect(timestring('5   D a YS    4 h   2 0     mI  nS')).to.equal(447600)
  })

  it('can parse a time string containing a decimal value', () => {
    expect(timestring('1.5 hours')).to.equal(5400)
    expect(timestring('2.75 mins')).to.equal(165)
    expect(timestring('1.5 w')).to.equal(907200)
  })

  it('can parse a number', () => {
    expect(timestring(300000)).to.equal(300)
    expect(timestring(5000.006)).to.equal(5)
    expect(timestring('600000', 'm')).to.equal(10)
  })
})


test/timestring.js
var chai = require('chai');
var expect = chai.expect;

var timestring = require('../build/timestring');

describe('timestring', function() {
  it('can parse a timestring', function() {
    var ts = new timestring();

    expect(ts.parse('1s')).to.equal(1);
    expect(ts.parse('1m')).to.equal(60);
    expect(ts.parse('1h')).to.equal(3600);
    expect(ts.parse('1d')).to.equal(86400);
    expect(ts.parse('1w')).to.equal(604800);
    expect(ts.parse('1mth')).to.equal(2419200);
    expect(ts.parse('1y')).to.equal(29030400);
  });

  it('can return a value in a specified unit', function() {
    expect((new timestring()).parse('5m', 's')).to.equal(300);
    expect((new timestring()).parse('5m', 'm')).to.equal(5);
  });

  it('uses the passed settings instead of the defaults', function() {
    var settings = {
      hoursPerDay: 1,
      daysPerWeek: 2,
      weeksPerMonth: 3,
      monthsPerYear: 4
    };

    var ts = new timestring(settings);

    expect(ts.parse('1d', 'h')).to.equal(1);
    expect(ts.parse('1w', 'd')).to.equal(2);
    expect(ts.parse('1mth', 'w')).to.equal(3);
    expect(ts.parse('1y', 'mth')).to.equal(4);
  });

  it('throws an error when an invalid unit is used in the timestring', function() {
    var ts = new timestring();

    expect(ts.parse.bind(ts, '1g')).to.throw(Error);
  });

  it('can parse a messy time string', function() {
    expect((new timestring()).parse('5   D a YS    4 h   2 0     mI  nS')).to.equal(447600);
  });

  it('should expose a method on String.prototype that will parse the string as a timestring', function(){
    var str = '1min';

    // no arguments passed
    expect(str.parseTime()).to.equal(60);

    // units argument passed
    expect(str.parseTime('m')).to.equal(1);

    // units + settings argument passed
    str = '5h';
    expect(str.parseTime('d', { hoursPerDay: 5 })).to.equal(1);
  });
});


Gulpfile.js
var gulp = require('gulp');
var $ = require('gulp-load-plugins')({
  scope: 'devDependencies'
});

var browserify = require('browserify');
var del = require('del');
var runSequence = require('run-sequence');
var source = require('vinyl-source-stream');

gulp.task('static-analysis:lint', function () {
  gulp.src('./src/**/*.js')
    .pipe($.jshint('.jshintrc'))
    .pipe($.jshint.reporter('jshint-stylish'));
});

gulp.task('static-analysis:cs', function () {
  return gulp.src('./src/**/*.js')
    .pipe($.jscs());
});

gulp.task('test', function () {
  return gulp.src('./test/timestring.js', { read: false })
    .pipe($.mocha());
});

gulp.task('build:clean', function (callback) {
  del(['./build/**/*'], callback);
});

gulp.task('build:transpile', function () {
  return gulp.src(['./src/Timestring.js', './src/String.parseTime.js'])
    .pipe($.sourcemaps.init())
    .pipe($.concat('timestring.js'))
    .pipe($.babel())
    .pipe($.sourcemaps.write())
    .pipe(gulp.dest('./build'));
});

gulp.task('build:lib', function () {
  return gulp.src('./build/timestring.js')
    .pipe(gulp.dest('./dist/lib'));
});

gulp.task('build:browserify', function() {
  return browserify('./build/timestring.js', { standalone: 'Timestring', debug: true })
    .bundle()
    .on('error', function(e){
        console.log(e.message);

        this.emit('end');
    })
    .pipe(source('timestring.js'))
    .pipe(gulp.dest('dist'));
});

gulp.task('build:minify', function () {
  return gulp.src('./dist/timestring.js')
    .pipe($.sourcemaps.init({ loadMaps: true }))
    .pipe($.uglify())
    .pipe($.rename({
      extname: '.min.js'
    }))
    .pipe($.sourcemaps.write())
    .pipe(gulp.dest('dist'));
});

gulp.task('build', function (callback) {
  runSequence(
    ['static-analysis:lint', 'static-analysis:cs'],
    'build:clean',
    'build:transpile',
    'test',
    ['build:lib', 'build:browserify'],
    'build:minify',
    callback
  );
});

gulp.task('sa', ['static-analysis:lint', 'static-analysis:cs']);

gulp.task('ci', function (callback) {
  runSequence(
    ['static-analysis:lint', 'static-analysis:cs'],
    'build:clean',
    'build:transpile',
    'test'
  );
});

gulp.task('default', ['build']);


Gruntfile.js
module.exports = function(grunt) {
  // measure the time each task takes
  require('time-grunt')(grunt);

  // autoload Grunt tasks
  require('load-grunt-tasks')(grunt);

  // main project config
  grunt.initConfig({
    pkg: grunt.file.readJSON('package.json'),
    uglify: {
      dist: {
        files: {
          'dist/<%= pkg.name %>.min.js': 'src/<%= pkg.name %>.js'
        }
      }
    },
    jshint: {
      options: {
        jshintrc: true
      },
      files: [
        'Gruntfile.js',
        'src/<%= pkg.name %>.js',
        'test/**/*.js'
      ]
    },
    mochaTest: {
      test: {
        options: {
          reporter: 'spec'
        },
        src: ['test/**/*.js']
      }
    }
  });

  // user defined tasks
  grunt.registerTask('test', ['mochaTest']);
  grunt.registerTask('lint', ['jshint']);
  grunt.registerTask('build', ['uglify']);
  grunt.registerTask('default', ['jshint', 'mochaTest', 'uglify']);
};


src/Timestring.js
export default class Timestring {
  constructor( settings = {} ) {
    const DEFAULT_SETTINGS = {
      hoursPerDay: 24,
      daysPerWeek: 7,
      weeksPerMonth: 4,
      monthsPerYear: 12
    };

    // merge default settings with user settings
    this.settings = DEFAULT_SETTINGS;
    for ( let s in settings ) { this.settings[s] = settings[s]; }

    // time units
    this.units = {
      s: [ "s", "sec", "secs", "second", "seconds" ],
      m: [ "m", "min", "mins", "minute", "minutes" ],
      h: [ "h", "hr", "hrs", "hour", "hours" ],
      d: [ "d", "day", "days" ],
      w: [ "w", "week", "weeks" ],
      mth: [ "mth", "mths", "month", "months" ],
      y: [ "y", "yr", "yrs", "year", "years" ]
    };

    // time unit seconds mappings
    this.unitValues = {
      s: 1,
      m: 60,
      h: 3600
    };

    // dynamic time unit seconds mappings
    // these are dynamic based on the settings
    this.unitValues.d = this.settings.hoursPerDay * this.unitValues.h;
    this.unitValues.w = this.settings.daysPerWeek * this.unitValues.d;
    this.unitValues.mth = this.settings.weeksPerMonth * this.unitValues.w;
    this.unitValues.y = this.settings.monthsPerYear * this.unitValues.mth;
  }

  parse( string, returnUnit = "s" ) {

    // get unit key helper
    let getUnitKey = ( unit ) => {
      for ( var k in this.units ) {
        for ( var u in this.units[k] ) {
          if ( unit === this.units[k][u] ) {
            return k;
          }
        }
      }

      // throw error if invalid unit was passed
      throw new Error( `The unit ${unit} is not supported by timestring` );
    };

    // convert a value to a specific unit
    let convert = ( value, unit ) => {
      let baseValue = this.unitValues[getUnitKey( unit )];

      return value / baseValue;
    };

    // get a value in seconds based on a specific unit
    let getSeconds = ( value, unit ) => {
      let baseValue = this.unitValues[getUnitKey( unit )];

      return value * baseValue;
    };

    // seconds counter
    let totalSeconds = 0;

    // split string into groups and get total seconds for each group
    let groups = string
                  .toLowerCase() // convert words to lower case
                  .replace( /[^\.\w+-]+/g, "" ) // remove white space
                  .match( /[-+]?[0-9]+[a-z]+/g ); // match time groups (digit followed by time unit - i.e 5d 15m = 2 time groups)

    if ( groups !== null ) {
      for ( let group in groups ) {
        let g = groups[group];
        let value = g.match( /[0-9]+/g )[0];
        let unit = g.match( /[a-z]+/g )[0];

        totalSeconds += getSeconds( value, unit );
      }
    }

    return convert( totalSeconds, returnUnit );
  }
}


src/timestring.js
export default class Timestring {
  constructor(settings = {}) {
    const DEFAULT_SETTINGS = {
      hoursPerDay: 24,
      daysPerWeek: 7,
      weeksPerMonth: 4,
      monthsPerYear: 12
    };

    // merge default settings with user settings
    this.settings = DEFAULT_SETTINGS;
    for (let s in settings) { this.settings[s] = settings[s]; }

    // time units
    this.units = {
      s: ['s', 'sec', 'secs', 'second', 'seconds'],
      m: ['m', 'min', 'mins', 'minute', 'minutes'],
      h: ['h', 'hr', 'hrs', 'hour', 'hours'],
      d: ['d', 'day', 'days'],
      w: ['w', 'week', 'weeks'],
      mth: ['mth', 'mths','month', 'months'],
      y: ['y', 'yr', 'yrs', 'year', 'years']
    };

    // time unit seconds mappings
    this.unitValues = {
      s: 1,
      m: 60,
      h: 3600
    };

    // dynamic time unit seconds mappings
    // these are dynamic based on the settings
    this.unitValues.d = this.settings.hoursPerDay * this.unitValues.h;
    this.unitValues.w = this.settings.daysPerWeek * this.unitValues.d;
    this.unitValues.mth = this.settings.weeksPerMonth * this.unitValues.w;
    this.unitValues.y = this.settings.monthsPerYear * this.unitValues.mth;
  }

  parse(string, returnUnit = 's') {
    // get unit key helper
    let getUnitKey = (unit) => {
      for (var k in this.units) {
        for (var u in this.units[k]) {
          if (unit === this.units[k][u]) {
            return k;
          }
        }
      }

      // throw error if invalid unit was passed
      throw new Error('The unit [' + unit + '] is not supported by timestring');
    };

    // convert a value to a specific unit
    let convert = (value, unit) => {
      let baseValue = this.unitValues[getUnitKey(unit)];

      return value / baseValue;
    };

    // get a value in seconds based on a specific unit
    let getSeconds = (value, unit) => {
      let baseValue = this.unitValues[getUnitKey(unit)];

      return value * baseValue;
    };

    // seconds counter
    let totalSeconds = 0;

    // split string into groups and get total seconds for each group
    let groups = string
                  .toLowerCase() // convert words to lower case
                  .replace(/[^\.\w+-]+/g, '') // remove white space
                  .match(/[-+]?[0-9]+[a-z]+/g); // match time groups (digit followed by time unit - i.e 5d 15m = 2 time groups)

    if (groups !== null) {
      for(let group in groups) {
        let g = groups[group];
        let value = g.match(/[0-9]+/g)[0];
        let unit = g.match(/[a-z]+/g)[0];

        totalSeconds += getSeconds(value, unit);
      }
    }

    return convert(totalSeconds, returnUnit);
  }
}


index.js
/**
 * Exports
 */

module.exports = parseTimestring

/**
 * Default options to use when parsing a timestring
 *
 * @type {Object}
 */

const DEFAULT_OPTS = {
  hoursPerDay: 24,
  daysPerWeek: 7,
  weeksPerMonth: 4,
  monthsPerYear: 12,
  daysPerYear: 365.25
}

/**
 * Map of accepted strings to unit
 *
 * @type {Object}
 */

const UNIT_MAP = {
  ms: ['ms', 'milli', 'millisecond', 'milliseconds'],
  s: ['s', 'sec', 'secs', 'second', 'seconds'],
  m: ['m', 'min', 'mins', 'minute', 'minutes'],
  h: ['h', 'hr', 'hrs', 'hour', 'hours'],
  d: ['d', 'day', 'days'],
  w: ['w', 'week', 'weeks'],
  mth: ['mon', 'mth', 'mths', 'month', 'months'],
  y: ['y', 'yr', 'yrs', 'year', 'years']
}

/**
 * Parse a timestring
 *
 * @param   {string|number} value
 * @param   {string} returnUnit
 * @param   {Object} opts
 * @returns {number}
 */

function parseTimestring (value, returnUnit, opts) {
  opts = Object.assign({}, DEFAULT_OPTS, opts || {})

  if (typeof value === 'number' || value.match(/^[-+]?[0-9.]+$/g)) {
    value = parseInt(value) + 'ms'
  }

  let totalSeconds = 0
  const unitValues = getUnitValues(opts)
  const groups = value
    .toLowerCase()
    .replace(/[^.\w+-]+/g, '')
    .match(/[-+]?[0-9.]+[a-z]+/g)

  if (groups === null) {
    throw new Error(`The value [${value}] could not be parsed by timestring`)
  }

  groups.forEach(group => {
    const value = group.match(/[0-9.]+/g)[0]
    const unit = group.match(/[a-z]+/g)[0]

    totalSeconds += getSeconds(value, unit, unitValues)
  })

  if (returnUnit) {
    return convert(totalSeconds, returnUnit, unitValues)
  }

  return totalSeconds
}

/**
 * Get unit values based on the passed options
 *
 * @param   {Object} opts
 * @returns {Object}
 */

function getUnitValues (opts) {
  const unitValues = {
    ms: 0.001,
    s: 1,
    m: 60,
    h: 3600
  }

  unitValues.d = opts.hoursPerDay * unitValues.h
  unitValues.w = opts.daysPerWeek * unitValues.d
  unitValues.mth = (opts.daysPerYear / opts.monthsPerYear) * unitValues.d
  unitValues.y = opts.daysPerYear * unitValues.d

  return unitValues
}

/**
 * Get the key for a unit
 *
 * @param   {string} unit
 * @returns {string}
 */

function getUnitKey (unit) {
  for (const key of Object.keys(UNIT_MAP)) {
    if (UNIT_MAP[key].indexOf(unit) > -1) {
      return key
    }
  }

  throw new Error(`The unit [${unit}] is not supported by timestring`)
}

/**
 *  Get the number of seconds for a value, based on the unit
 *
 * @param   {number} value
 * @param   {string} unit
 * @param   {Object} unitValues
 * @returns {number}
 */

function getSeconds (value, unit, unitValues) {
  return value * unitValues[getUnitKey(unit)]
}

/**
 * Convert a value from its existing unit to a new unit
 *
 * @param   {number} value
 * @param   {string} unit
 * @param   {Object} unitValues
 * @returns {number}
 */

function convert (value, unit, unitValues) {
  return value / unitValues[getUnitKey(unit)]
}
